### ISSUE TITLE
[OP] TargetEncoding

### ISSUE THREAD

**Issue by bschifferer** [NVIDIA/NVTabular#115]
**Is your feature request related to a problem? Please describe.**
TargetEncoding is a common feature engineering for categorical variables in tabular datasets. In RecSys2020, it was one of the most used techniques in our pipeline. It would be great if the target column is a parameter in the OP. 

**Describe the solution you'd like**
We just consider the case of a numerical target or binary classification problem. For each categorical value, the mean of the target is calculated and a new feature (column) is created with the means of the categorical value. To prevent overfitting, to additional logics are applied:
1. Training CrossValidation: To prevent overfitting in the training dataset, a cross validation strategy is applied to the training set. The training set is splitted in k folds and the means for the i-th fold is calculated with all reamining folds. The fold assignmed is randomly. The validation/test set uses the full training dataset for calculating the mean.
2. Smoothing: To prevent overfitting for low cardinality categories, the means a smoothed with the overall mean of the target variable.

Function:
TE = ((mean_cat*count_cat)+(mean_global*p_smooth)) / (count_cat+p_smooth)

count_cat := count of the categorical value
mean_cat := mean of target value for the categorical value 
mean_global := mean of the target value in the dataset
p_smooth := smoothing factor

The OP should support multi-column Target Encoding. For Example, if the dataset contains Language and Gender, we can target-encode the combination pairs of ['Language', 'Gender']. The number of columns should be not limited to 2.
The OP has two parameters: 
* kfold := how many folds to apply on the training dataset 
* p_smooth := smoothing factor

**Describe alternatives you've considered**
My idea would be to create an op and define GroupByMoments as a statistic.
GroupByMoments has to be extended to support kfold operation.
The OP needs to differentiate between applying to the trainingset (using kfold) and validation/testset (using all folds)

**Additional context**
An example implementation can be found here:
https://github.com/rapidsai/recsysChallenge2020/blob/master/Baselines/DASK-XGB/Demo-e2e-multi-CPU.ipynb : MTE_one_shot
https://github.com/rapidsai/recsysChallenge2020/blob/master/final_models/XGB1334/XGB1334-submit-singleGPU.ipynb : target_encode_cudf_v3

**Comment by benfred** [comment]
We should add the ability to do folds to the groupby op to implement this

**Comment by rjzamora** [comment]
I reviewed the RecSys 2020 notebook and it looks like we can take a similar approach in NVTabular.  One distinction is that we will probably want to use the logic behind our (more-scalable) `CategoryStatistics` op (rather than the upstream dask-groupby API).  Here are the changes  I think we will need:

- **Start by addressing #121** - Having primitives for multi-column groupby aggregations will be very convenient here.
- We need a way to specify the “fold” for each row in the dataset.  The easiest approach is the one used in the RecSys notebook: We assign a new “fold” column (or `”__fold__”` to establish its private nature) for every row in the dataset.  The challenge with this approach is that we are effectively requiring a feature-engineering operation (add random integer column) as a dependency for both the `TargetEncoding` operation and its `CategoryStatistics` dependency.
    - It is possible to avoid the addition of a new column to the global Dask DataFrame if we can uniquely map each row of the dataset to a fold index.  **If we are comfortable mapping entire partitions onto the same fold index, this is easy**.   Unless I am misunderstanding, the RecSys notebook is actually doing something similar to this anyway (the same fold index is always assigned to a large consecutive group of rows).  If we are allowed to assign fold indices to each partition in a round-robin fashion, we can avoid adding a `”__fold__”` column to the global dask_cudf DataFrame (and just do it temporarily within each partition-wise task)
- **Assuming we can map entire partitions onto folds** (or accept an existing `fold_column` from the user), we should be able to simply use `CategoryStatistics` op (previously called `BroupByMoments`) to get the count/sum statistics of our target column(s) for both an isolated `cat_column` and a `[cat_column, ”__fold__”]` combination.   Within `TargetEncoding`, the worker process can simply combine (merge) these statistics, and subtract out the fold-based statistics from the total statistics for each `[cat_column, ”__fold__”]`-category.  This result can be cached (only performed once on each worker) and the apply logic will simply calculate the  target encoding from these statistics, and add the TE column.  At apply time, we should know the fold index for each row from a user-specified  `fold_column` or by taking a modulus of the partition index.

**Comment by bschifferer** [comment]
I don't think you have often more folds than partitions. I think the number of folds are in the range (3,10).

If the dataset is shuffled for the partitions, then it is possible to use the partitions for the folds. The counter example, the dataset is sorted by userid and I want to TargetEncode the userid. Then it is pretty likely that all user with userid=0 are in the first partition. Then the outoffold encoding will not work as all other partitions have no information for userid=0.

**Comment by rjzamora** [comment]
> I don't think you have often more folds than partitions. I think the number of folds are in the range [1,10].

Right - You can always force the number of DataFrame partitions to be the desired amount, but I agree that the "performant" partition count for small/modest datasets will be a problem.

> The counter example, the dataset is sorted by userid and I want to TargetEncode the userid. Then it is pretty likely that all user with userid=0 are in the first partition. Then the outoffold encoding will not work as all other partitions have no information for userid=0.

Totally agree - This is exactly the kind of case where we would probably need to add a global `"__fold__"` column (or expect the user to add one).  Originally, I did not consider it a possibility to use the DataFrame partition as a proxy for fold index for this reason. However, the [RecSys2020 notebook](https://github.com/rapidsai/recsysChallenge2020/blob/master/Baselines/DASK-XGB/Demo-e2e-multi-CPU.ipynb) made is seem like this *could* work in some cases.
